src/backend/src/services/slack.services.ts

```diff
@@ -0,0 +1,289 @@
+   * @param organizationId the id of the organization represented by the slack api
+   * @returns an annoucement if an announcement was processed and created/modified/deleted
+   */
+  static async processMessageSent(event: SlackMessageEvent, organizationId: string): Promise<Announcement | undefined> {
+    let slackChannelName: string;
+    //get the name of the channel from the slack api
+    try {
+      slackChannelName = await getChannelName(event.channel);
+    } catch (error) {
+      slackChannelName = `Unknown_Channel:${event.channel}`;
+    }
+    const dateCreated = new Date(Number(event.event_ts));
+
+    //get the message that will be processed either as the event or within a subtype
+    let eventMessage: SlackMessage;
+
+    if (event.subtype) {
+      switch (event.subtype) {
+        case 'message_deleted':
+          //delete the message using the client_msg_id
+          eventMessage = (event as SlackDeletedMessage).previous_message;
+          try {
+            return AnnouncementService.deleteAnnouncement(eventMessage.client_msg_id, organizationId);
+          } catch (ignored) {
+            return;
+          }
+        case 'message_changed':
+          eventMessage = (event as SlackUpdatedMessage).message;
+          break;
+        default:
+          //other events that do not effect announcements
+          return;
+      }
+    } else {
+      eventMessage = event as SlackMessage;
+    }
+
+    //loop through the blocks of the meta data while accumulating the
+    //text and users notified
+    let messageText = '';
+    let userIdsToNotify: string[] = [];
+
+    //Get the settings of all users in this organization to compare slack ids
+    const users = await UsersService.getAllUsers();
+    const userSettings = await Promise.all(
+      users.map((user) => {
+        return UsersService.getUserSettings(user.userId);
+      })
+    );
+
+    //get the name of the user that sent the message from slack
+    let userName: string = '';
+    try {
+      userName = (await getUserName(eventMessage.user)) ?? '';
+    } catch (ignored) {}
+
+    //if slack could not produce the name of the user, look for their name in prisma
+    if (!userName) {
+      const userIdList = userSettings
+        .filter((userSetting) => userSetting.slackId === eventMessage.user)
+        .map((userSettings) => userSettings.userId);
+      if (userIdList.length !== 0) {
+        const prismaUserName = users.find((user) => user.userId === userIdList[0]);
+        userName = prismaUserName
+          ? `${prismaUserName?.firstName} ${prismaUserName?.lastName}`
+          : 'Unknown User:' + eventMessage.user;
+      } else {
+        userName = 'Unknown_User:' + eventMessage.user;
+      }
+    }
+
+    //pull out the blocks of data from the metadata within the message event
+    const richTextBlocks = eventMessage.blocks?.filter((eventBlock: any) => eventBlock.type === 'rich_text');
+
+    if (richTextBlocks && richTextBlocks.length === 1) {
```

Comment:
why does this have to be exactly one?

Reply:
The events can have multiple blocks in them but they would only ever have one with type 'rich_text' so this test is just making sure that one exists. We could change it to > 0 but in theory it would never matter

Reply:
so if i did @channel and @here it would only have one rich text?

Reply:
yes, any message will have one rich_text block, this will then contain within it an array of elements representing the different blocks in the message. Its a bit confusing (and kind of poorly designed if you ask me), but every message has a rich_text block which has a rich_text_section within it which has the elements that we need to parse.

Reply:
ok sounds good